0.27 : compilation is broken with NLS option enabled under MinGW

compilation trace : 

/mnt/devel/GIT/6.x/project/bundles/mxe/temp.build/ext_exiv2/ext_exiv2-prefix/src/ext_exiv2/src/futils.cpp: In function 'std::__cxx11::string Exiv2::getProcessPath()':
/mnt/devel/GIT/6.x/project/bundles/mxe/temp.build/ext_exiv2/ext_exiv2-prefix/src/ext_exiv2/src/futils.cpp:467:9: error: 'HANDLE' was not declared in this scope
         HANDLE processHandle = NULL;
         ^
/mnt/devel/GIT/6.x/project/bundles/mxe/temp.build/ext_exiv2/ext_exiv2-prefix/src/ext_exiv2/src/futils.cpp:468:9: error: 'processHandle' was not declared in this scope
         processHandle = OpenProcess(PROCESS_QUERY_INFORMATION | PROCESS_VM_READ, FALSE, GetCurrentProcessId());
         ^
/mnt/devel/GIT/6.x/project/bundles/mxe/temp.build/ext_exiv2/ext_exiv2-prefix/src/ext_exiv2/src/futils.cpp:468:37: error: 'PROCESS_QUERY_INFORMATION' was not declared in this scope
         processHandle = OpenProcess(PROCESS_QUERY_INFORMATION | PROCESS_VM_READ, FALSE, GetCurrentProcessId());
                                     ^
/mnt/devel/GIT/6.x/project/bundles/mxe/temp.build/ext_exiv2/ext_exiv2-prefix/src/ext_exiv2/src/futils.cpp:468:65: error: 'PROCESS_VM_READ' was not declared in this scope
         processHandle = OpenProcess(PROCESS_QUERY_INFORMATION | PROCESS_VM_READ, FALSE, GetCurrentProcessId());
                                                                 ^
/mnt/devel/GIT/6.x/project/bundles/mxe/temp.build/ext_exiv2/ext_exiv2-prefix/src/ext_exiv2/src/futils.cpp:468:82: error: 'FALSE' was not declared in this scope
         processHandle = OpenProcess(PROCESS_QUERY_INFORMATION | PROCESS_VM_READ, FALSE, GetCurrentProcessId());
                                                                                  ^
/mnt/devel/GIT/6.x/project/bundles/mxe/temp.build/ext_exiv2/ext_exiv2-prefix/src/ext_exiv2/src/futils.cpp:468:109: error: 'GetCurrentProcessId' was not declared in this scope
         processHandle = OpenProcess(PROCESS_QUERY_INFORMATION | PROCESS_VM_READ, FALSE, GetCurrentProcessId());
                                                                                                             ^
/mnt/devel/GIT/6.x/project/bundles/mxe/temp.build/ext_exiv2/ext_exiv2-prefix/src/ext_exiv2/src/futils.cpp:468:110: error: 'OpenProcess' was not declared in this scope
         processHandle = OpenProcess(PROCESS_QUERY_INFORMATION | PROCESS_VM_READ, FALSE, GetCurrentProcessId());
                                                                                                              ^
/mnt/devel/GIT/6.x/project/bundles/mxe/temp.build/ext_exiv2/ext_exiv2-prefix/src/ext_exiv2/src/futils.cpp:470:13: error: 'TCHAR' was not declared in this scope
             TCHAR filename[MAX_PATH];
             ^
/mnt/devel/GIT/6.x/project/bundles/mxe/temp.build/ext_exiv2/ext_exiv2-prefix/src/ext_exiv2/src/futils.cpp:471:58: error: 'filename' was not declared in this scope
             if (GetModuleFileNameEx(processHandle, NULL, filename, MAX_PATH) != 0) {
                                                          ^
/mnt/devel/GIT/6.x/project/bundles/mxe/temp.build/ext_exiv2/ext_exiv2-prefix/src/ext_exiv2/src/futils.cpp:471:68: error: 'MAX_PATH' was not declared in this scope
             if (GetModuleFileNameEx(processHandle, NULL, filename, MAX_PATH) != 0) {
                                                                    ^
/mnt/devel/GIT/6.x/project/bundles/mxe/temp.build/ext_exiv2/ext_exiv2-prefix/src/ext_exiv2/src/futils.cpp:471:76: error: 'GetModuleFileNameEx' was not declared in this scope
             if (GetModuleFileNameEx(processHandle, NULL, filename, MAX_PATH) != 0) {
                                                                            ^
/mnt/devel/GIT/6.x/project/bundles/mxe/temp.build/ext_exiv2/ext_exiv2-prefix/src/ext_exiv2/src/futils.cpp:474:38: error: 'CloseHandle' was not declared in this scope
             CloseHandle(processHandle);
                                      ^
src/CMakeFiles/exiv2lib.dir/build.make:363 : la recette pour la cible « src/CMakeFiles/exiv2lib.dir/futils.cpp.obj » a échouée
gmake[6]: *** [src/CMakeFiles/exiv2lib.dir/futils.cpp.obj] Erreur 1
gmake[6]: *** Attente des tâches non terminées....
/mnt/devel/GIT/6.x/project/bundles/mxe/temp.build/ext_exiv2/ext_exiv2-prefix/src/ext_exiv2/src/convert.cpp:1510:10: warning: 'bool {anonymous}::convertStringCharsetWindows(std::__cxx11::string&, const char*, const char*)' defined but not used [-Wunused-function]
     bool convertStringCharsetWindows(std::string& str, const char* from, const char* to)
          ^
CMakeFiles/Makefile2:212 : la recette pour la cible « src/CMakeFiles/exiv2lib.dir/all » a échouée
gmake[5]: *** [src/CMakeFiles/exiv2lib.dir/all] Erreur 2
Makefile:129 : la recette pour la cible « all » a échouée
gmake[4]: *** [all] Erreur 2
ext_exiv2/CMakeFiles/ext_exiv2.dir/build.make:112 : la recette pour la cible « ext_exiv2/ext_exiv2-prefix/src/ext_exiv2-stamp/ext_exiv2-build » a échouée
gmake[3]: *** [ext_exiv2/ext_exiv2-prefix/src/ext_exiv2-stamp/ext_exiv2-build] Erreur 2
CMakeFiles/Makefile2:415 : la recette pour la cible « ext_exiv2/CMakeFiles/ext_exiv2.dir/all » a échouée
gmake[2]: *** [ext_exiv2/CMakeFiles/ext_exiv2.dir/all] Erreur 2
CMakeFiles/Makefile2:427 : la recette pour la cible « ext_exiv2/CMakeFiles/ext_exiv2.dir/rule » a échouée
gmake[1]: *** [ext_exiv2/CMakeFiles/ext_exiv2.dir/rule] Erreur 2
Makefile:196 : la recette pour la cible « ext_exiv2 » a échouée
gmake: *** [ext_exiv2] Erreur 2
FAILED COMMAND: ${MXE_BUILD_TARGETS}-cmake --build . --config RelWithDebInfo --target ext_exiv2 -- -j$CPU_CORES

[gilles@localhost mxe]$ 

In fact, in futils.cpp there are 2 problems in getProcessPath() : 

    std::string getProcessPath()
    {
        std::string ret("unknown");
    #if defined(WIN32)                                       <======= Must be _MSC_VER
        HANDLE processHandle = NULL;
        processHandle = OpenProcess(PROCESS_QUERY_INFORMATION | PROCESS_VM_READ, FALSE, GetCurrentProcessId());
        if (processHandle != NULL) {
            TCHAR filename[MAX_PATH];
            if (GetModuleFileNameEx(processHandle, NULL, filename, MAX_PATH) != 0) {
                ret = filename;
            }
            CloseHandle(processHandle);
        }
    #elif defined(__APPLE__)
        const int pid = getpid();
        char pathbuf[PROC_PIDPATHINFO_MAXSIZE];
        if (proc_pidpath (pid, pathbuf, sizeof(pathbuf)) > 0) {
            ret = pathbuf;
        }
    #elif defined(__linux__) || defined(__CYGWIN__) || defined(__MINGW__)
        // http://stackoverflow.com/questions/606041/how-do-i-get-the-path-of-a-process-in-unix-linux
        char proc[100];
        char path[500];
        sprintf(proc,"/proc/%d/exe", getpid());
        ssize_t l = readlink (proc, path,sizeof(path)-1);     << =========== Do not exists under Mingw
        if (l>0) {
            path[l]=0;
            ret = path;
        }
    #endif
    #if defined(WIN32)
        const size_t idxLastSeparator = ret.find_last_of('\\');
    #else
        const size_t idxLastSeparator = ret.find_last_of('/');
    #endif
        return ret.substr(0, idxLastSeparator);
    }

// ---------------------------------------

I take a look under Qt5 code which is fully portable and support very well MinGW, the readlink() call is wrapped with access()

Gilles Caulier

If i replace : 

ssize_t l = readlink (proc, path,sizeof(path)-1);

by : 
ssize_t l = 0; //readlink (proc, path,sizeof(path)-1);

All the core library compile fine.

[gilles@localhost build]$ make
[ 20%] Built target xmp
[ 48%] Built target exiv2lib_int
Scanning dependencies of target exiv2lib
[ 49%] Building CXX object src/CMakeFiles/exiv2lib.dir/futils.cpp.obj
[ 50%] Building CXX object src/CMakeFiles/exiv2lib.dir/image.cpp.obj
/mnt/devel/GIT/6.x/project/bundles/mxe/temp.build/ext_exiv2/ext_exiv2-prefix/src/ext_exiv2/src/image.cpp:871:77: warning: unused parameter 'useCurl' [-Wunused-parameter]
     BasicIo::AutoPtr ImageFactory::createIo(const std::wstring& wpath, bool useCurl)
                                                                             ^
[ 51%] Building CXX object src/CMakeFiles/exiv2lib.dir/ini.cpp.obj
[ 52%] Building CXX object src/CMakeFiles/exiv2lib.dir/iptc.cpp.obj
[ 54%] Building CXX object src/CMakeFiles/exiv2lib.dir/jp2image.cpp.obj
[ 55%] Building CXX object src/CMakeFiles/exiv2lib.dir/jpgimage.cpp.obj
[ 56%] Building CXX object src/CMakeFiles/exiv2lib.dir/matroskavideo.cpp.obj
[ 57%] Building CXX object src/CMakeFiles/exiv2lib.dir/metadatum.cpp.obj
[ 58%] Building CXX object src/CMakeFiles/exiv2lib.dir/mrwimage.cpp.obj
[ 60%] Building CXX object src/CMakeFiles/exiv2lib.dir/orfimage.cpp.obj
[ 61%] Building CXX object src/CMakeFiles/exiv2lib.dir/pgfimage.cpp.obj
[ 62%] Building CXX object src/CMakeFiles/exiv2lib.dir/preview.cpp.obj
[ 63%] Building CXX object src/CMakeFiles/exiv2lib.dir/properties.cpp.obj
[ 64%] Building CXX object src/CMakeFiles/exiv2lib.dir/psdimage.cpp.obj
[ 65%] Building CXX object src/CMakeFiles/exiv2lib.dir/quicktimevideo.cpp.obj
[ 67%] Building CXX object src/CMakeFiles/exiv2lib.dir/rafimage.cpp.obj
[ 68%] Building CXX object src/CMakeFiles/exiv2lib.dir/riffvideo.cpp.obj
[ 69%] Building CXX object src/CMakeFiles/exiv2lib.dir/rw2image.cpp.obj
[ 70%] Building CXX object src/CMakeFiles/exiv2lib.dir/tags.cpp.obj
[ 71%] Building CXX object src/CMakeFiles/exiv2lib.dir/tgaimage.cpp.obj
[ 72%] Building CXX object src/CMakeFiles/exiv2lib.dir/tiffimage.cpp.obj
[ 74%] Building CXX object src/CMakeFiles/exiv2lib.dir/types.cpp.obj
[ 75%] Building CXX object src/CMakeFiles/exiv2lib.dir/utilsvideo.cpp.obj
[ 76%] Building CXX object src/CMakeFiles/exiv2lib.dir/value.cpp.obj
[ 77%] Building CXX object src/CMakeFiles/exiv2lib.dir/version.cpp.obj
[ 78%] Building CXX object src/CMakeFiles/exiv2lib.dir/webpimage.cpp.obj
[ 80%] Building CXX object src/CMakeFiles/exiv2lib.dir/xmp.cpp.obj
[ 81%] Building CXX object src/CMakeFiles/exiv2lib.dir/xmpsidecar.cpp.obj
[ 82%] Building CXX object src/CMakeFiles/exiv2lib.dir/pngimage.cpp.obj
[ 83%] Linking CXX shared library ../bin/libexiv2.dll
[100%] Built target exiv2lib

So the problem is only located to found a right replacement of readlink() call under Mingw, or to disable this function for the moment under MinGW.

As i can see, readlink() support is problematic, as symbolic link is completly different under Windows.

https://stackoverflow.com/questions/39785170/replacement-for-readlink-function-in-windows

Gilles

After another few hours of building, MXE dropped dead having not compiled a single line of Exiv2 code.

```
Makefile:796: recipe for target '/media/psf/Home/gnu/github/exiv2/exiv2/build.win32/usr/i686-w64-mingw32.static/installed/nettle' failed
make: *** [/media/psf/Home/gnu/github/exiv2/exiv2/build.win32/usr/i686-w64-mingw32.static/installed/nettle] Error 1
rmills@rmillsmm-ubuntu:/media/psf/Home/gnu/github/exiv2/exiv2/build.win32$ cmake --build . --target help
Error: could not load cache
rmills@rmillsmm-ubuntu:/media/psf/Home/gnu/github/exiv2/exiv2/build.win32$ cmake --build .
Error: could not load cache
rmills@rmillsmm-ubuntu:/media/psf/Home/gnu/github/exiv2/exiv2/build.win32$ make
[build]       nettle                 i686-w64-mingw32.static
[done]        nettle                 i686-w64-mingw32.static                                  0 KiB          0m34.772s
[download]    gnutls-3.6.5.tar.xz
[build]       gnutls                 i686-w64-mingw32.static

Failed to build package gnutls for target i686-w64-mingw32.static!
------------------------------------------------------------
guile/src/Makefile.am:117: warning: '%'-style pattern rules are a GNU make extension
configure: WARNING: unrecognized options: --enable-gtk-doc, --enable-gtk-doc-html, --enable-gtk-doc-pdf, --with-html-dir, --disable-doxygen
configure: error: pwd does not report name of working directory
Makefile:796: recipe for target 'build-only-gnutls_i686-w64-mingw32.static' failed
make[1]: *** [build-only-gnutls_i686-w64-mingw32.static] Error 1
make[1]: Leaving directory '/media/psf/Home/gnu/github/exiv2/exiv2/build.win32'
real	0m50.163s
user	0m32.980s
sys	0m6.118s
------------------------------------------------------------
[log]      /media/psf/Home/gnu/github/exiv2/exiv2/build.win32/log/gnutls_i686-w64-mingw32.static

Makefile:796: recipe for target '/media/psf/Home/gnu/github/exiv2/exiv2/build.win32/usr/i686-w64-mingw32.static/installed/gnutls' failed
make: *** [/media/psf/Home/gnu/github/exiv2/exiv2/build.win32/usr/i686-w64-mingw32.static/installed/gnutls] Error 1
rmills@rmillsmm-ubuntu:/media/psf/Home/gnu/github/exiv2/exiv2/build.win32$ 
```

I don't intend to spend further time on this issue.  The supported method of building Exiv2 for MinGW is to use msys2 as documented in README.md.

--- a/.gitlab-ci.yml
+++ b/.gitlab-ci.yml
@@ -36,6 +36,14 @@ Fedora:
   <<: *default_config
   <<: *distro_build
 
+Fedora_MinGW:
+  image: fedora:29
+  before_script:
+    - dnf -y upgrade
+    - dnf -y install mingw64-gcc-c++ mingw64-filesystem mingw64-expat mingw64-zlib cmake make
+  script:
+    - python3 ci/test_build.py --without-tests --cmake-executable "mingw64-cmake" --cmake-options "-DEXIV2_TEAM_EXTRA_WARNINGS=ON -DEXIV2_ENABLE_VIDEO=ON -DEXIV2_ENABLE_WEBREADY=ON -DEXIV2_ENABLE_WIN_UNICODE=ON " --compilers --shared-libs OFF
+
 Debian:
   image: debian:9
   <<: *default_config

--- a/ci/test_build.py
+++ b/ci/test_build.py
@@ -8,23 +8,6 @@ import shlex
 import subprocess
 import sys
 
-#: -DEXIV2_BUILD_SHARED_LIBS options
-SHARED_LIBS = ["ON", "OFF"]
-
-#: C & C++ compiler as tuples
-CCS = ["gcc", "clang"]
-
-#: -DCMAKE_BUILD_TYPE options
-BUILD_TYPES = ["Debug", "Release"]
-
-#: Additional parameters for cmake
-CMAKE_OPTIONS = os.getenv("CMAKE_OPTIONS") or \
-    "-DEXIV2_TEAM_EXTRA_WARNINGS=ON -DEXIV2_ENABLE_WEBREADY=ON "\
-	"-DEXIV2_BUILD_UNIT_TESTS=ON -DBUILD_WITH_CCACHE=ON "
-
-#: cpu count
-NCPUS = multiprocessing.cpu_count()
-
 
 def call_wrapper(*args, **kwargs):
     """
@@ -36,41 +19,114 @@ def call_wrapper(*args, **kwargs):
         sys.exit(return_code)
 
 
-# create build directory
-os.mkdir("build")
+def matrix_build(shared_libs, ccs, build_types, cmake_bin, cmake_options,
+                 tests=True):
 
-root_dir = os.path.abspath(os.getcwd())
+    NCPUS = multiprocessing.cpu_count()
 
-for params in itertools.product(SHARED_LIBS, CCS, BUILD_TYPES):
+    os.mkdir("build")
 
-    lib_type, cc, build_type = params
+    for params in itertools.product(shared_libs, ccs, build_types):
 
-    cxx = {"gcc": "g++", "clang": "clang++"}[cc]
+        lib_type, cc, build_type = params
 
-    cwd = os.path.abspath(os.path.join("build", "_".join(params)))
-    os.mkdir(cwd)
+        cwd = os.path.abspath(
+            os.path.join(
+                "build",
+                "_".join(
+                    map(lambda p: str(p) if p is not None else "", params)
+                )
+            )
+        )
+        os.mkdir(cwd)
 
-    cmake = "cmake {!s} -DCMAKE_BUILD_TYPE={build_type} "\
-        "-DBUILD_SHARED_LIBS={lib_type} ../.."\
-        .format(CMAKE_OPTIONS, build_type=build_type, lib_type=lib_type)
-    make = "make -j " + str(NCPUS)
-    make_tests = "make tests"
-    unit_tests = os.path.join(cwd, "bin", "unit_tests")
+        cmake = "{cmake_bin} {!s} -DCMAKE_BUILD_TYPE={build_type} " \
+            "-DBUILD_SHARED_LIBS={lib_type} -DEXIV2_BUILD_UNIT_TESTS={tests} "\
+            "../..".format(
+                cmake_options, cmake_bin=cmake_bin, build_type=build_type,
+                lib_type=lib_type, tests="ON" if tests else "OFF"
+            )
+        make = "make -j " + str(NCPUS)
+        make_tests = "make tests"
+        unit_test_binary = os.path.join(cwd, "bin", "unit_tests")
 
-    # set up environment
-    env_copy = os.environ.copy()
-    env_copy["CC"] = cc
-    env_copy["CXX"] = cxx
+        # set compiler via environment only when requested
+        env_copy = os.environ.copy()
+        if cc is not None:
+            cxx = {"gcc": "g++", "clang": "clang++"}[cc]
+            env_copy["CC"] = cc
+            env_copy["CXX"] = cxx
 
-    # location of the binaries for the new test suite:
-    env_copy["EXIV2_BINDIR"] = os.path.join(cwd, "bin")
+        # location of the binaries for the new test suite:
+        env_copy["EXIV2_BINDIR"] = os.path.join(cwd, "bin")
 
-    kwargs = {"env": env_copy, "cwd": cwd}
+        kwargs = {"env": env_copy, "cwd": cwd}
 
-    def run(cmd):
-        call_wrapper(shlex.split(cmd), **kwargs)
+        def run(cmd):
+            call_wrapper(shlex.split(cmd), **kwargs)
 
-    run(cmake)
-    run(make)
-    run(make_tests)
-    run(unit_tests)
+        run(cmake)
+        run(make)
+        if tests:
+            run(make_tests)
+            run(unit_test_binary)
+
+
+if __name__ == '__main__':
+    import argparse
+
+    parser = argparse.ArgumentParser(
+        description="Build and test exiv2 using a matrix of build switches")
+    parser.add_argument(
+        "--compilers",
+        help="Compilers to be used to build exiv2 (when none ore specified, "
+        "then the default compiler will be used)",
+        nargs='*',
+        default=["gcc", "clang"],
+        type=str
+    )
+    parser.add_argument(
+        "--shared-libs",
+        help="Values for the -DBUILD_SHARED_LIBS option",
+        nargs='+',
+        default=["ON", "OFF"],
+        type=str
+    )
+    parser.add_argument(
+        "--build-types",
+        help="Values for the -DCMAKE_BUILD_TYPE option",
+        nargs='+',
+        default=["Debug", "Release"],
+        type=str
+    )
+    parser.add_argument(
+        "--cmake-executable",
+        help="alternative name or path for the cmake executable",
+        nargs=1,
+        default=['cmake'],
+        type=str
+    )
+    parser.add_argument(
+        "--without-tests",
+        help="Skip building and running tests",
+        action='store_true'
+    )
+    parser.add_argument(
+        "--cmake-options",
+        help="Additional flags for cmake",
+        type=str,
+        nargs='?',
+        default="-DEXIV2_TEAM_EXTRA_WARNINGS=ON -DEXIV2_ENABLE_WEBREADY=ON"
+        " -DBUILD_WITH_CCACHE=ON -DEXIV2_ENABLE_CURL=ON"
+    )
+
+    args = parser.parse_args()
+
+    if len(args.compilers) == 0:
+        args.compilers = [None]
+
+    matrix_build(
+        args.shared_libs, args.compilers, args.build_types,
+        args.cmake_executable[0], args.cmake_options,
+        not args.without_tests
+    )

--- a/src/futils.cpp
+++ b/src/futils.cpp
@@ -17,12 +17,6 @@
  * along with this program; if not, write to the Free Software
  * Foundation, Inc., 51 Franklin Street, 5th Floor, Boston, MA 02110-1301 USA.
  */
-/*
-  File:      futils.cpp
-  Author(s): Andreas Huggel (ahu) <ahuggel@gmx.net>
-  History:   08-Dec-03, ahu: created
-             02-Apr-05, ahu: moved to Exiv2 namespace
- */
 // *****************************************************************************
 // included header files
 #include "config.h"
@@ -33,25 +27,26 @@
 // + standard includes
 #include <sys/types.h>
 #include <sys/stat.h>
-
-#ifdef _MSC_VER
-    #include <Windows.h>
-    # define S_ISREG(m)      (((m) & S_IFMT) == S_IFREG)
-    #include <psapi.h>  // For access to GetModuleFileNameEx
-#elif defined(__APPLE__)
-    #include <libproc.h>
-#endif
-
-#ifdef EXV_HAVE_UNISTD_H
-  # include <unistd.h>                     // for stat()
-#endif
-
 #include <cstdio>
 #include <cerrno>
 #include <sstream>
 #include <cstring>
 #include <algorithm>
 #include <stdexcept>
+#ifdef   EXV_HAVE_UNISTD_H
+#include <unistd.h>                     // for stat()
+#endif
+
+#if defined(WIN32)
+#include <windows.h>
+#include <psapi.h>  // For access to GetModuleFileNameEx
+#endif
+
+#if defined(_MSC_VER)
+#define S_ISREG(m)      (((m) & S_IFMT) == S_IFREG)
+#elif defined(__APPLE__)
+#include <libproc.h>
+#endif
 
 namespace Exiv2 {
     const char* ENVARDEF[] = {"/exiv2.php", "40"}; //!< @brief default URL for http exiv2 handler and time-out
@@ -468,7 +463,7 @@ namespace Exiv2 {
         if (proc_pidpath (pid, pathbuf, sizeof(pathbuf)) > 0) {
             ret = pathbuf;
         }
-    #elif defined(__linux__) || defined(__CYGWIN__) || defined(__MINGW__)
+    #elif defined(__linux__) || defined(__CYGWIN__) || defined(__MSYS__)
         // http://stackoverflow.com/questions/606041/how-do-i-get-the-path-of-a-process-in-unix-linux
         char proc[100];
         char path[500];
@@ -479,11 +474,8 @@ namespace Exiv2 {
             ret = path;
         }
     #endif
-    #if defined(WIN32)
-        const size_t idxLastSeparator = ret.find_last_of('\\');
-    #else
-        const size_t idxLastSeparator = ret.find_last_of('/');
-    #endif
+
+        const size_t idxLastSeparator = ret.find_last_of(EXV_SEPARATOR_STR);
         return ret.substr(0, idxLastSeparator);
     }
 }                                       // namespace Exiv2

